Add New Blockstore-related Opaque Key Types

bradenmacdonald opened the pull request:
This PR moves some key-related code for new Blockstore-based content out of `edx-platform` and into `opaque-keys`.

It includes: 
* A new abstract key type, `LearningContextKey`, which identifies a _learning context_. A learning context may be a course, a library, or other types in the future.
    - All `CourseKey`s are `LearningContextKey`s
* A new `LibraryLocatorV2` key type which is a type of `LearningContextKey` that identifies a Blockstore-based content library.
    - Note that unlike the existing modulestore-based [`LibraryLocator`](https://github.com/edx/opaque-keys/blob/72e54d1d40aa8209f3eef5238f64794f5804895f/opaque_keys/edx/locator.py#L378), calling `isinstance(LibraryLocatorV2, Coursekey)` will return `False` because we now have this more abstract type to use. Yay.
* A new `LibraryUsageLocatorV2` key type which identifies a usage of an XBlock in a Blockstore-based content library.
* A new concrete `DefinitionKey` type, `BundleDefinitionLocator`. These keys are mostly used internally in the new Blockstore-based XBlock runtime. Any XBlock in Blockstore will have one of these keys as its `.scope_ids.def_id` value. A `BundleDefinitionLocator` specifies a specific XBlock in a specific OLX file in a specific version (or draft) of a specific Blockstore bundle.

## Implications for developers

This PR has one change that affects existing code: it changes the `KEY_TYPE` of `CourseKey` and `CourseLocator` to `context_key`. Which impacts `ccx-keys` and requires a corresponding change there: https://github.com/edx/ccx-keys/pull/8  . Both that change and this must be brought into edx-platform at the same time, or some CCX-related tests will fail.

Also, as we roll out and use more of the new content library code in the platform, it will be the case that other parts of the LMS and related apps may start seeing XBlock UsageKeys that refer to Blockstore-based content libraries. (So far the new runtime is completely isolated but coming PRs will change that.) The `.context_key` property of any usage key will return the `LearningContextKey` (usually a CourseKey). But, importantly, so will the `.course_key` property, which may produce unexpected results if the code in question doesn't know what do to with a non-course key.

Because the old `LibraryLocator` classes inherit from `CourseKey`, dealing with all this could get a little messy, so I've added a nice `.is_course` property to all `LearningContextKey` types that will always give you the correct answer. So there are now two recommended patterns for reading usage keys if your code only knows how to deal with courses:

If your code should work with any course, but you don't want to deal with libraries or other things:
```
from opaque_keys.edx.keys import UsageKey
usage_key = UsageKey.from_string(...whatever...)
if not usage_key.context_key.is_course:
    return  # Ignore this entry if this code only knows how to deal with courses. It could be a modulestore-based content library or blockstore-based content library, or some new thing
# proceed with processing usage_key
```

If your code is pretty specific to modulestore-based courses and you think it might break with future Blockstore-based courses:
```
from opaque_keys.edx.keys import UsageKey
from opaque_keys.edx.locator import CourseLocator
usage_key = UsageKey.from_string(...whatever...)
if not isinstance(usage_key.context_key, CourseLocator):
    return  # Ignore anything that's not a modulestore course
# proceed with processing usage_key
```

Same goes if you have a generic `context_key` or `course_key`: you should use either `context_key.is_course` or `isinstance(course_key, CourseLocator)` (depending on if your code is modulestore-specific or not) and ignore the key when these return `False`.

Likewise when parsing a key, use `LearningContextKey.from_string(...)` if you don't care if it's a course or not, and use `CourseKey.from_string(...)` if you expected a course. In either case, wrap that in a `try: ... except InvalidKeyError: ...` so you can handle unexpected key types gracefully.

ormsbee commented on setup.py:
```diff
@@ -31,15 +31,17 @@
             'dict = opaque_keys.tests.test_opaque_keys:DictKey',
         ],
-        'course_key': [
+        'context_key': [
```
Yeah, ok, this is fine. Thank you.

bradenmacdonald commented:
Squashed for merge! As noted, once this v1.1.0 is deployed on PyPI, we need to update + merge the `ccx-keys` PR and then the `edx-platform` PR. If anyone pulls this new version 1.1.0 into edx-platform before those other PRs are merged, it will cause tests to fail.

bradenmacdonald commented:
Updated to version 2.0 since the entry point change from `course_key` to `context_key` is backwards-incompatible and causes the CCX keys (which are in a separate package) to be not found by edx-platform until they are also updated.

edx-webhook commented:
Although this pull request is already merged, I've created [OSPR-5414](https://openedx.atlassian.net/browse/OSPR-5414) so that we can track it in Jira.

There is nothing you have to do. No action is needed from your side.  Thanks again for your contribution.

<!-- comment:welcome_closed -->
<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->
